Refactor tests

janosbrodbeck opened the pull request:
Sorry that it turned out so big and really hard to review now. Was not intended :D In case of uncertainty, let me know, we can also go over together.

* Switch from `testing.postgresql` to `pytest-postgresql`
* Use pytest fixtures for creating the database and getting the AsyncClient (http client)
* Use async tests (necessary now)
* Place all tests in `app/tests`
* `utils/models.py` provides some reusable functions to create data in the database
* Testdata is placed in `tests/testdata` and contains functions to get the data and to validate
* Contains some fix commits. Not clean, but I stumbled across them while writing tests, and it is easier to justify them when directly covered by tests and working
* **GitHub actions already uses the new tests in this PR**

Currently covered from tests (Sonar Cloud code coverage integration planned as a follow-up):
- All CRUD operations
- Most API endpoints (only the multi get currently missing), even codegen and project upload
- CodeGeneration
- PandasParser with Excel, CSV and JSON

`conftest.py` contains the fixtures for Pytest, which are injected in the parameters of the test functions. The scope of the database is currently per test, so each test gets a whole new database. A cross-test database would be possible too, if necessary.

`pytestmark = pytest.mark.asyncio` at the top of some files marks all tests as `@pytest.mark.asyncio` (async tests) automatically.

dolevi-tub commented on src/app/tests/api/v1/test_welding_point.py:
```diff
@@ -0,0 +1,105 @@
+    assert response.status_code == 200
+
+    # GET endpoint returns a list of welding points associated to a project
+    assert len(response.json()) == 0
```
Please still check whether the database has zero entries too. That's even more important than the response body

janosbrodbeck replied:
Addressed

dolevi-tub commented on src/app/tests/utils/utils.py:
```diff
@@ -0,0 +1,26 @@
+
+
+def random_float(negative=True) -> float:
+    if negative:
+        return random.uniform(-5.0, 5.0)
+    else:
+        return random.uniform(0.0, 5.0)
```
```suggestion
    return random.uniform(-5.0 if negative else 0.0, 5.0)
```
Or does this look weird?

dolevi-tub commented on src/app/tests/pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
+asyncio_mode = strict
```
Please add an explanatory comment on why we need this mode

janosbrodbeck replied:
I added a comment, at least in what it does. Also made it more lax with `auto`. I don't remove the `pytest.asyncio` in the files as I feel that it is cleaner. It allows us theoretically to use other async test frameworks, when the mode is strict. But I am quite sure we won't need another. The annotations are practically a leftover, but see no reason to remove them.

dolevi-tub commented on src/app/tests/crud/test_crud_welding_point.py:
```diff
@@ -0,0 +1,90 @@
+    project_obj = await create_project(db=database)
+
+    welding_point_obj = await create_welding_point(db=database, project_obj=project_obj)
+    tmp = await welding_point.get_by_id(db=database, id=welding_point_obj.id)
+    assert welding_point_obj.__eq__(tmp)
```
Rename this to something mroe intuitive than `tmp`.

janosbrodbeck replied:
Addressed

dolevi-tub commented on src/app/tests/crud/test_crud_robot_type.py:
```diff
@@ -0,0 +1,47 @@
+
+async def test_crud_robot_type_get(database: AsyncSession):
+    robot_type_obj = await create_robot_type(db=database)
+    tmp = await robot_type.get_by_id(db=database, id=robot_type_obj.id)
+    assert robot_type_obj.__eq__(tmp)
```
Please use a more accesible name than `tmp`.

janosbrodbeck replied:
Addressed

dolevi-tub commented on src/app/tests/crud/test_crud_robot.py:
```diff
@@ -0,0 +1,66 @@
+
+    robot_obj = await create_robot(db=database, robot_type_obj=robot_type_obj)
+    assert robot_obj is not None
+    assert isinstance(robot_obj, Robot)
```
Please also assert the correctness of the `robot_type_id` of the `robot_obj`.

And: Please also check the that the object has arrived in the database. (And not just that I return the request body and still get a pass :wink:)

janosbrodbeck replied:
Addressed

dolevi-tub commented on src/app/tests/crud/test_crud_robot.py:
```diff
@@ -0,0 +1,66 @@
+    robot_type_obj = await create_robot_type(db=database)
+
+    robot_obj = await create_robot(db=database, robot_type_obj=robot_type_obj)
+    tmp = await robot.get_by_id(db=database, id=robot_obj.id)
+    assert robot_obj.__eq__(tmp)
```
Please use a more accessible variable name than `tmp`.

janosbrodbeck replied:
Addressed

dolevi-tub commented on src/app/tests/crud/test_crud_robot.py:
```diff
@@ -0,0 +1,66 @@
+
+    robot_obj = await create_robot(db=database, robot_type_obj=robot_type_obj)
+    await robot.update(db=database, db_obj=robot_obj, obj_in={"description": "modified"})
+    assert robot_obj.description == "modified"
```
Please also assert that the id has not changed.

janosbrodbeck replied:
It is allowed to change the ID right now. Actually, I wanted to prevent this by UpdateTypes not having the ID, but unfortunately they automatically inherit that from the base class. We need to have another look at the topic again, where this can be prevented best. But I would rather see this as a follow-up.

dolevi-tub commented on src/app/tests/crud/test_crud_robot.py:
```diff
@@ -0,0 +1,66 @@
+    assert robot_obj.description == "modified"
+
+    await robot.update(db=database, db_obj=robot_obj, obj_in=RobotUpdate(description="modified again"))
+    assert robot_obj.description == "modified again"
```
Please also assert that the id has not changed.

janosbrodbeck replied:
See other comment

dolevi-tub commented on src/app/tests/crud/test_crud_robot_type.py:
```diff
@@ -0,0 +1,47 @@
+async def test_crud_robot_type_update(database: AsyncSession):
+    robot_type_obj = await create_robot_type(db=database)
+    await robot_type.update(db=database, db_obj=robot_type_obj, obj_in={"name": "modified"})
+    assert robot_type_obj.name == "modified"
```
Please also assert that the id has not changed.

janosbrodbeck replied:
See other comment

dolevi-tub commented on src/app/tests/crud/test_crud_robot_type.py:
```diff
@@ -0,0 +1,47 @@
+    assert robot_type_obj.name == "modified"
+
+    await robot_type.update(db=database, db_obj=robot_type_obj, obj_in=RobotTypeUpdate(name="modified again"))
+    assert robot_type_obj.name == "modified again"
```
Please also assert that the id has not changed.

janosbrodbeck replied:
See other comment

dolevi-tub commented on src/app/tests/crud/test_crud_welding_point.py:
```diff
@@ -0,0 +1,90 @@
+
+    welding_point_obj = await create_welding_point(db=database, project_obj=project_obj)
+    await welding_point.update(db=database, db_obj=welding_point_obj, obj_in={"name": "modified"})
+    assert welding_point_obj.name == "modified"
```
Please also assert that the id has not changed.

janosbrodbeck replied:
See other comment

dolevi-tub commented on src/app/tests/crud/test_crud_welding_point.py:
```diff
@@ -0,0 +1,90 @@
+    assert welding_point_obj.name == "modified"
+
+    await welding_point.update(db=database, db_obj=welding_point_obj, obj_in=WeldingPointUpdate(name="modified again"))
+    assert welding_point_obj.name == "modified again"
```
Please also assert that the id has not changed.

janosbrodbeck replied:
See other comment

dolevi-tub commented on src/app/tests/crud/test_crud_project.py:
```diff
@@ -0,0 +1,47 @@
+async def test_crud_project_update(database: AsyncSession):
+    project_obj = await create_project(db=database)
+    await project.update(db=database, db_obj=project_obj, obj_in={"name": "modified"})
+    assert project_obj.name == "modified"
```
Please also assert that the id has not changed.

dolevi-tub commented on src/app/tests/crud/test_crud_project.py:
```diff
@@ -0,0 +1,47 @@
+    assert project_obj.name == "modified"
+
+    await project.update(db=database, db_obj=project_obj, obj_in=ProjectUpdate(name="modified again"))
+    assert project_obj.name == "modified again"
```
Please also assert that the id has not changed.

janosbrodbeck replied:
See other comment

dolevi-tub commented on src/app/tests/crud/test_crud_project.py:
```diff
@@ -0,0 +1,47 @@
+
+async def test_crud_project_get(database: AsyncSession):
+    project_obj = await create_project(db=database)
+    tmp = await project.get_by_id(db=database, id=project_obj.id)
+    assert project_obj.__eq__(tmp)
```
Please use a more accessible variable name than `tmp`

dolevi-tub commented on src/app/tests/crud/test_crud_generation_template.py:
```diff
@@ -0,0 +1,48 @@
+
+    await generation_template.update(db=database, db_obj=generation_template_obj,
+                                     obj_in=GenerationTemplateUpdate(name="modified again"))
+    assert generation_template_obj.name == "modified again"
```
Please also assert that the id has not changed.

dolevi-tub replied:
See other comments

dolevi-tub commented on src/app/tests/crud/test_crud_generation_template.py:
```diff
@@ -0,0 +1,48 @@
+
+async def test_crud_generation_template_get(database: AsyncSession):
+    generation_template_obj = await create_generation_template(db=database)
+    tmp = await generation_template.get_by_id(db=database, id=generation_template_obj.id)
+    assert generation_template_obj.__eq__(tmp)
```
Please use a more accessible variable name than `tmp`

janosbrodbeck replied:
More or less addressed

dolevi-tub commented on src/app/tests/crud/test_crud_generation_template.py:
```diff
@@ -0,0 +1,48 @@
+async def test_crud_generation_template_create(database: AsyncSession):
+    generation_template_obj = await create_generation_template(db=database)
+    assert generation_template_obj is not None
+    assert isinstance(generation_template_obj, GenerationTemplate)
```
Please also check the that the object has arrived in the database. (And not just that I return the request body and still get a pass :wink:)

janosbrodbeck replied:
Addressed

dolevi-tub commented on src/app/tests/crud/test_crud_project.py:
```diff
@@ -0,0 +1,47 @@
+async def test_crud_project_create(database: AsyncSession):
+    project_obj = await create_project(db=database)
+    assert project_obj is not None
+    assert isinstance(project_obj, Project)
```
Please also check the that the object has arrived in the database. (And not just that I return the request body and still get a pass :wink:)

janosbrodbeck replied:
Addressed

dolevi-tub commented on src/app/tests/crud/test_crud_robot_type.py:
```diff
@@ -0,0 +1,47 @@
+async def test_crud_robot_type_create(database: AsyncSession):
+    robot_type_obj = await create_robot_type(db=database)
+    assert robot_type_obj is not None
+    assert isinstance(robot_type_obj, RobotType)
```
Please also check the that the object has arrived in the database. (And not just that I return the request body and still get a pass :wink:)

janosbrodbeck replied:
Addressed

dolevi-tub commented on src/app/tests/crud/test_crud_welding_point.py:
```diff
@@ -0,0 +1,90 @@
+
+    welding_point_obj = await create_welding_point(db=database, project_obj=project_obj)
+    assert welding_point_obj is not None
+    assert isinstance(welding_point_obj, WeldingPoint)
```
Please also check the that the object has arrived in the database. (And not just that I return the request body and still get a pass :wink:)

janosbrodbeck replied:
Addressed

dolevi-tub reviewed (changes requested):
Great PR overall. I love it :D
I think most important is also asserting the database state on creation tests and that the id has not changed on update.

Anyways, this gets three thumbs up by me: :+1: :+1: :+1:

janosbrodbeck commented on src/app/tests/api/v1/test_codegen.py:
```diff
@@ -0,0 +1,26 @@
+    assert response.status_code == 200
+
+    content = response.json()
+    assert content == f"{welding_point_obj.x}, {welding_point_obj.y}, {welding_point_obj.z} / " \
+                      f"{welding_point_obj.roll}, {welding_point_obj.pitch}, {welding_point_obj.yaw} / " \
+                      f"{welding_point_obj.welding_order}"
```
Yes and no. `unittest` and `pytest` are two distinct testing frameworks. It is basically possible to use unittest with pytest, but according to my research it is not possible to use fixtures in class methods (class methods would be the tests) directly. We would have to make a special fixture within the class to get the fixtures or mark the whole class to get the fixtures. But then the database and client would be used for the whole class, what I don't want to do. I like that we get for each test a whole new database. It would be possible, but we would need more boilerplate, which I personally would try to avoid except when we really gain something from. 

In my opinion, the current error output is sufficient for our case and I do not see the use in us to write more boilerplate for a little more info. For real applications it would make sense to look again at this topic :D

We could also print out e.g. the response every time as a trade-off. The print output is only displayed on error.

dolevi-tub replied:
Okay if it's not feasible it's not feasible :D

janosbrodbeck commented on src/app/tests/api/v1/test_generation_template.py:
```diff
@@ -0,0 +1,55 @@
+    assert content["content"] == generation_template_obj.content
+
+
+async def test_delete_robot_type(client: AsyncClient, database: AsyncSession):
+    generation_template_obj = await create_generation_template(db=database)
+    await client.delete(f"{settings.API_V1_STR}/generationtemplate/:id?_id={generation_template_obj.id}")
+    response = await client.get(f"{settings.API_V1_STR}/generationtemplate/:id?_id={generation_template_obj.id}")
+    assert response.status_code == 404
```
Isn't this basically a happy case test? I would just add the comparison of the result. But a dictionary test would always fail here:
`generation_template_obj` is of type `GenerationTemplate` which is a pydantic model class, containing more than the result. E.g. it also contains the `config` key, which we configured for our schemas. So we would already have to exclude these keys from the comparison. Its sadly not that simple.

janosbrodbeck commented on src/app/tests/api/v1/test_welding_point.py:
```diff
@@ -0,0 +1,105 @@
+pytestmark = pytest.mark.asyncio
+
+
+async def test_create_welding_point_fail(client: AsyncClient, database: AsyncSession):
```
Addressed

janosbrodbeck commented on src/app/tests/utils/models.py:
```diff
@@ -0,0 +1,71 @@
+from app.schemas.robot import RobotCreate
+from app.schemas.robot_type import RobotTypeCreate
+from app.schemas.welding_point import WeldingPointCreate
+from app.tests.utils.utils import random_string, random_float, get_example_template
```
Oops. I even looked for a module doing that when writing these utility functions. Totally forgot about it and thought I had done that already.

janosbrodbeck commented:
I addressed some of your comments.
* API: Delete tests now checks, whether the object was really deleted in db
* API: Dedicated tests for `404` when requesting not existing data. Moved this out of the delete test, as it was not correct here
* API: Checks in create tests, if object is also created in db
* CRUD: Check, whether objects are created in the database. Also compared to the returned object
* Random data: Now uses the seed set as pytest parameter

janosbrodbeck commented on src/app/tests/api/v1/test_project.py:
```diff
@@ -0,0 +1,71 @@
+    project_obj = await create_project(db=database)
+    await client.delete(f"{settings.API_V1_STR}/project/:id?_id={project_obj.id}")
+    response = await client.get(f"{settings.API_V1_STR}/project/:id?_id={project_obj.id}")
+    assert response.status_code == 404
```
We compare the created and returned object now and check, whether the object in the database was really deleted

janosbrodbeck commented on src/app/tests/api/v1/test_robot_type.py:
```diff
@@ -0,0 +1,53 @@
+    robot_type_obj = await create_robot_type(db=database)
+    await client.delete(f"{settings.API_V1_STR}/robottype/:id?_id={robot_type_obj.id}")
+    response = await client.get(f"{settings.API_V1_STR}/robottype/:id?_id={robot_type_obj.id}")
+    assert response.status_code == 404
```
We compare the created and returned object now and check, whether the object in the database was really deleted

janosbrodbeck commented on .github/workflows/docker-image.yml:
```diff
@@ -34,7 +34,7 @@ jobs:
       - name: Run tests
         working-directory: ./src/
-        run: python3 -m pytest -vv --durations=0
+        run: python3 -m pytest app/tests
```
That was a stupid mistake on my part. Its back!

dolevi-tub commented on src/app/tests/utils/utils.py:
```diff
@@ -0,0 +1,26 @@
+    if negative:
+        return random.randint(-5000, 5000)
+    else:
+        return random.randint(0, 5000)
```
Good resolution :D

dolevi-tub reviewed (approved):
Merge it please :smile:
